libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp

```diff
@@ -0,0 +1,286 @@
+    if (uart != nullptr) {
+        uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance));
+    }
+}
+
+// detect if a serial port has been setup to accept rangefinder input
+bool AP_RangeFinder_BLPing::detect(AP_SerialManager &serial_manager, uint8_t serial_instance)
+{
+    return serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance) != nullptr;
+}
+
+/*
+   update the state of the sensor
+*/
+void AP_RangeFinder_BLPing::update(void)
+{
+    uint32_t now = AP_HAL::millis();
+
+    if (get_reading(state.distance_cm)) {
+        // update range_valid state based on distance measured
+        state.last_reading_ms = AP_HAL::millis();
+        update_status();
+    } else if (now - state.last_reading_ms > BLPING_TIMEOUT_MS) {
+        set_status(RangeFinder::RangeFinder_NoData);
+
+        // initialise sensor if no distances recently
+        if (now - last_init_ms > BLPING_INIT_RATE_MS) {
+            last_init_ms = now;
+            init_sensor();
+        }
+    }
+}
+
+// initialise sensor
+void AP_RangeFinder_BLPing::init_sensor()
+{
+    // request distance from sensor
+    send_message(BLDPIN_MSGID_DISTANCE_SIMPLE, nullptr, 0);
+}
+
+// send message to sensor
+void AP_RangeFinder_BLPing::send_message(uint16_t msgid, uint8_t *payload, uint16_t payload_len)
+{
+    if (uart == nullptr) {
+        return;
+    }
+
+    // write header
+    uart->write((uint8_t)BLPING_FRAME_HEADER1);
+    uart->write((uint8_t)BLPING_FRAME_HEADER2);
+    uint16_t crc = BLPING_FRAME_HEADER1 + BLPING_FRAME_HEADER2;
+
+    // write payload length
+    uart->write(LOWBYTE(payload_len));
+    uart->write(HIGHBYTE(payload_len));
+    crc += LOWBYTE(payload_len) + HIGHBYTE(payload_len);
+
+    // msgid
+    uart->write(LOWBYTE(msgid));
+    uart->write(HIGHBYTE(msgid));
+    crc += LOWBYTE(msgid) + HIGHBYTE(msgid);
+
+    // src dev id
+    uart->write((uint8_t)BLPING_SRC_ID);
+    crc += BLPING_SRC_ID;
+
+    // destination dev id
+    uart->write((uint8_t)BLPING_DEST_ID);
+    crc += BLPING_DEST_ID;
+
+    // payload
+    if (payload != nullptr) {
+        for (uint16_t i = 0; i<payload_len; i++) {
+            uart->write(payload[i]);
+            crc += payload[i];
+        }
+    }
+
+    // checksum
+    uart->write(LOWBYTE(crc));
+    uart->write(HIGHBYTE(crc));
+}
+
+// distance returned in reading_cm, signal_ok is set to true if sensor reports a strong signal
+bool AP_RangeFinder_BLPing::get_reading(uint16_t &reading_cm)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    float sum_cm = 0;
+    uint16_t count = 0;
+
+    // read any available lines from the lidar
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        int16_t b = uart->read();
+        if (b < 0) {
+            continue;
+        }
+        if (parse_byte(b)) {
+            count++;
+            sum_cm += distance_cm;
```

Comment:
How about using a return parameter here rather than storing `distance_cm` in `*this`?